javascript/ql/src/semmle/javascript/frameworks/jQuery.qll

```diff
@@ -258,26 +258,145 @@ private class JQueryChainedElement extends DOM::Element, InvokeExpr {
 }
 
 /**
- * A model of a URL request made using the `jQuery.ajax` or `jQuery.getJSON`.
+ * A model of a URL request made using the `jQuery.ajax`.
  */
-private class JQueryClientRequest extends ClientRequest::Range {
-  JQueryClientRequest() {
-    exists(string name |
-      name = "ajax" or
-      name = "getJSON"
-    |
-      this = jquery().getAMemberCall(name)
-    )
-  }
+private class JQueryAjaxCall extends ClientRequest::Range {
+  JQueryAjaxCall() { this = jquery().getAMemberCall("ajax") }
 
   override DataFlow::Node getUrl() {
-    result = getArgument(0) or
+    result = getArgument(0) and not exists(getOptionArgument(0, _))
+    or
     result = getOptionArgument([0 .. 1], "url")
   }
 
   override DataFlow::Node getHost() { none() }
 
   override DataFlow::Node getADataNode() { result = getOptionArgument([0 .. 1], "data") }
+
+  private string getResponseType() {
+    getOptionArgument([0 .. 1], "dataType").mayHaveStringValue(result)
+  }
+
+  override DataFlow::Node getAResponseDataNode(string responseType, boolean promise) {
+    (
+      responseType = getResponseType()
+      or
+      not exists(getResponseType()) and responseType = ""
+    ) and
+    promise = false and
+    (
+      result =
+        getOptionArgument([0 .. 1], "success")
+            .getALocalSource()
+            .(DataFlow::FunctionNode)
+            .getParameter(0)
+      or
+      result =
+        getAResponseNodeFromAnXHRObject(getOptionArgument([0 .. 1],
+            any(string method | method = "error" or method = "complete"))
+              .getALocalSource()
+              .(DataFlow::FunctionNode)
+              .getParameter(0))
+      or
+      result = getAnAjaxCallbackDataNode(this)
+      or
+      result = getAResponseNodeFromAnXHRObject(getAnXHRObject(this))
+    )
+  }
+}
+
+/**
+ * Gets the response data node from a call to a jqXHR Object.
```

Comment:
```suggestion
 * Gets a response data node from a call to a jqXHR Object.
```
Alternatively: rename the predicate to `getAjaxCallbackDataNode`. I think that is what I prefer since you generally expect a single such node.

Reply:
Also: is there such a thing as `a call to a jqXHR Object`? Or do you mean something like `a call to a method on jqXHR Object`?